### DISCUSSION TITLE
Xiao NRF52840 board support with UF2 bootloader

### DISCUSSION THREAD

**Discussion by favnec5** [micropython/micropython#10127] [Hardware & Peripherals] [answered]
hello,
I try to install micropython on my fresh Xio NRF52840.
Circuitpython does it. So i reproduced the pins, compiled, converted to uf2 (like as suggested here : https://forum.micropython.org/viewtopic.php?f=12&t=7783). 
And ... nothing :-/
I don't understand. micropython doesn't start when the UF2 file is dropped.
Why this difference between circuitpython and MP ? 
I don't see what's wrong.
Does anyone have any ideas?
regards
Thomas

**Comment by robert-hh** [comment]
I just got a MP image running based on the pca100056 board with changed pin assignment. The REPL is at the UART pins. At the moment I load through SDIO/SDCLK with a Segger JLink adapter. At the moment it looks as if entering the bootloader via double reset is gone. And machine.bootloader() is just a dummy. More work to do.

**Comment by robert-hh** [comment]
So I managed to compile a package and load it to the XIAO nrd52840 board. Instructions below. The files are in the branch https://github.com/robert-hh/micropython/tree/nrf_xiao

---------------- instructions --------------------
Make and install a firmware package for SEEED nrf52840:

1. IInstall adafruit-nrfutil with `apt-get install adafruit-nrfutil` 
   The official NRF nrfutil suite used in Makefile did not work for me and    failed consistently with Python errors. Maybe a wrong Python version, but obviously it misses version checks.
2. Create a firmware for the SEEED nrf52840. to do so:
   - you have to use Softdevice v7.0.1 or v7.3.0. The 6.1.1. Version does
     not match the bootloader.
   - compile with `make BOARD=<board name> -j5`
     The mpconfigboard.mk has to define SD=s140 and SOFTDEV_VERSION=7.x.x.

3. Upload the firmware:

   - generate in the build directory the nrfutil package with:

     `nrfutil dfu genpkg --application firmware.hex --dev-type 82 firmware.zip`

   - Set the board to dfu mode by pushing reset twice fast, until the
     drive appears. Using machine.bootloader() or baud rate switch does not work (yet).

   - Upload the firmware through USB serial with:

     `nrfutil dfu serial -pkg firmware.zip --port /dev/ttyACM0 --baudrate 460800 -sb
`

Sample mpconfigboard.mk:

 ```
MCU_SERIES = m4
MCU_VARIANT = nrf52
MCU_SUB_VARIANT = nrf52840
SOFTDEV_VERSION = 7.3.0
SD=s140
LD_FILES += boards/nrf52840_1M_256k.ld
NRF_DEFINES += -DNRF52840_XXAA
MICROPY_VFS_LFS2 = 1
```
Sample s140_7.3.0.ld file:
```
/* GNU linker script for s140 SoftDevice version 7.3.0 */
_sd_size = 0x00027000;
_sd_ram  = 0x0000e000;
```

The S140 package v7.3.0 can be downloaded from the Nordic Semi site at:

https://www.nordicsemi.com/Products/Development-software/S140/Download

and has to be placed into nrf/drivers/bluetooth next to the other versions. Someone
may as well patch the download_ble_stack.sh script to do so. I just do not know
the file tags for the 7.3.0 versions.

**Comment by robert-hh** [comment, marked as answer]
@favnec5 So now there is some success in uploading .uf2 file to the upload drive. Instructions:

1. Download and install u2fconv.py

2. Create a firmware for the SEEED nrf52840. to do so:
   - you have to use Softdevice v7.0.1 or v7.3.0. The 6.1.1. Version does
     not match the bootloader.
   - compile with make BOARD=\<board name\> -j5
     The mpconfigboard.mk has to define SD=s140 and SOFTDEV_VERSION=7.x.x.

3. Create the .uf2 file with the command:

   `uf2conv.py -c -f 0xADA52840 -o firmware.uf2 firmware.hex`

   It must report that the start address is 0x27000. If you omit the -o option,
   the output file will have the name flash.uf2.

4. Enable the upload mode by double reset push or calling machine.bootloader()
   and copy the file firmware.uf2 to the board

The config files are the same as in my previous post. Just the final packaging of the .uf2 file is different.

**Reply by favnec5** [reply]
thank you  robert, it works like a charm !! :-)

**Reply by favnec5** [reply]
Just a question : why " -j5" i searched but not found what j5 specifies ?

**Reply by robert-hh** [reply]
The -jx option tells make to run in x parallel threads. That  speeds up the compile. If you have a 8 core CPU, you can use -j8 or -j9. Usins a too large number does not hurt.

**Reply by favnec5** [reply]
ok ! thanks

**Reply by robert-hh** [reply]
Another note: If you use the most recent version of my branch, you can call:

`make -j5 BOARD=SEEED_XIAO uf2`

That will create the uf2 file as part of the build process as firmware.uf2 in the build directory.

**Comment by robert-hh** [comment]
I tried also to use the internal flash for a file system with my nrf_xiao branch. It uses spiflash.py and flashbdev from this repository: https://github.com/robert-hh/SPI_Flash and the little mount script below. SPI(3) has to be used for the baudrate of 32MHz. The other SPI devices only support up to 8 MHz.
```
import os
from flashbdev import FlashBdev
from spiflash import SPIflash
from machine import SPI, Pin

wp = Pin("QSPI_D2", Pin.OUT, value=1)
hold = Pin("QSPI_D3", Pin.OUT, value=1)
cs = Pin("QSPI_CS", Pin.OUT, value=1)
spi = SPI(3, sck=Pin("QSPI_SCK"), mosi=Pin("QSPI_D0"), miso=Pin("QSPI_D1"), baudrate=32_000_000)
flash = FlashBdev(SPIflash(spi, cs))
try:
    vfs = os.VfsLfs2(flash, progsize=256)
except OSError as e:
    print("Mount failed with error", e)
    print("Recreate the file system")  # Previous content is lost!
    os.VfsLfs2.mkfs(flash, progsize=256)
    vfs = os.VfsLfs2(flash, progsize=256)

os.mount(vfs, "/data")
```

**Reply by favnec5** [reply]
Nice, 2040Kb That's more.
I'll check it out, but I have the impression that micropython consumes much less power than its brother at bootloader startup. And that's really cool.

**Reply by robert-hh** [reply]
The supply current see in REPL is like 5mA, which matches the data sheet. It will be higher during RF transmission (~10mA) and can get above that when driving external devices through GPIO pins.
The machine methods lightsleep(), deepsleep(), idle() and sleep() are more or less dummy functions, just executing either __WFE() or __WFI(). Which means they just return quite fast, as far as I could tell after at least one ms.

**Reply by doh38** [reply]
hi, 
I tried to use the nightly built and created a uf2 from it, also tried  to compile the firmware with a git clone today. Both boot and run well, but there seems to be no bluetooth module loaded. Is this because it's "too early" or there is an issue with the last sources?

**Reply by robert-hh** [reply]
The adaption for the XIAO board just add the board definitions, but is not different at the core features to the other nrf52840 boards. I seems that the bluetooth approach of the nrf board is different to the ESP32. There is a ble module and a ubluepy module related to bluetooth. What I miss too is the documentation.

**Reply by doh38** [reply]
ah, thanks. I definitively need to learn more. I come from the CircuitPython and thought it was similar. time to learn more about ble and ubluepy!
Ein schönen abend noch :)

**Comment by doh38** [comment]
Continuing my quest to get BLE uart working I succeeded to advertise the NUS with this "code":
```
from ubluepy import Service, Characteristic, UUID, Peripheral, constants
from machine import Pin
import time 

led_red = Pin('P26', Pin.OUT)
led_green = Pin('P30', Pin.OUT)
led_blue = Pin('P6', Pin.OUT)

p = Peripheral()

def event_handler(id, handle, data):
    global periph
    global service
    if id == constants.EVT_GAP_CONNECTED:
        print("connected")
        led_green.low()
    elif id == constants.EVT_GAP_DISCONNECTED:
        led_green.high()
        # restart advertisement
        print("restart advertisement")
        periph.advertise(device_name="XIAO_BLE", services=[service])
    elif id == constants.EVT_GATTS_WRITE:
        led_red.high() if int(data[0]) else led_red.low()


# start off with LED(1) off
led_red.high()



notif_enabled = False
uuid_service = UUID("6E400001-B5A3-F393-E0A9-E50E24DCCA9E")
uuid_uart_rx    = UUID("6E400002-B5A3-F393-E0A9-E50E24DCCA9E")
uuid_uart_tx    = UUID("6E400003-B5A3-F393-E0A9-E50E24DCCA9E")
service = Service(uuid_service)
char_rx = Characteristic(uuid_uart_rx, props=Characteristic.PROP_WRITE | Characteristic.PROP_WRITE_WO_RESP)
char_tx = Characteristic(uuid_uart_tx, props=Characteristic.PROP_READ | Characteristic.PROP_NOTIFY, attrs = Characteristic.ATTR_CCCD)

service.addCharacteristic(char_rx)
service.addCharacteristic(char_tx) 

periph = Peripheral()
periph.addService(service)
periph.setConnectionHandler(event_handler)
periph.advertise(device_name="XIAO_BLE", services=[service])


while (True):
    time.sleep_ms(500)
```
I can connect and the little green led will go on. 
But I absolutely don't see any ble_uart Service available anywhere. 

I have tried to activate BLE_NUS in the bluetooth conf.h like this *but I have no idea if this is what's missing and/or correct:
```

#elif (BLUETOOTH_SD == 140)

#define MICROPY_PY_BLE                  (1)
#define MICROPY_PY_BLE_NUS              (1)
#define BLUETOOTH_WEBBLUETOOTH_REPL     (0)
#define MICROPY_PY_UBLUEPY              (1)
#define MICROPY_PY_UBLUEPY_PERIPHERAL   (1)
#define MICROPY_PY_UBLUEPY_CENTRAL      (1)

#else
#error "SD not supported"
#endif

// Default defines.

#ifndef MICROPY_PY_BLE
#define MICROPY_PY_BLE                  (0)
#endif

#ifndef MICROPY_PY_BLE_NUS
#define MICROPY_PY_BLE_NUS              (0)
#endif

#endif
```
I then get a compilation following error:

```
CC build-seeed_xiao_nrf52-s140/pins_gen.c
CC build-seeed_xiao_nrf52-s140/frozen_content.c
LINK build-seeed_xiao_nrf52-s140/firmware.elf
build-seeed_xiao_nrf52-s140/drivers/usb/usb_cdc.o (symbol from plugin): In function `USBD_IRQHandler':
(.text+0x0): multiple definition of `mp_hal_stdio_poll'
build-seeed_xiao_nrf52-s140/drivers/bluetooth/ble_uart.o (symbol from plugin):(.text+0x0): first defined here
build-seeed_xiao_nrf52-s140/drivers/usb/usb_cdc.o (symbol from plugin): In function `USBD_IRQHandler':
(.text+0x0): multiple definition of `mp_hal_stdin_rx_chr'
build-seeed_xiao_nrf52-s140/drivers/bluetooth/ble_uart.o (symbol from plugin):(.text+0x0): first defined here
build-seeed_xiao_nrf52-s140/drivers/usb/usb_cdc.o (symbol from plugin): In function `USBD_IRQHandler':
(.text+0x0): multiple definition of `mp_hal_stdout_tx_strn'
build-seeed_xiao_nrf52-s140/drivers/bluetooth/ble_uart.o (symbol from plugin):(.text+0x0): first defined here
build-seeed_xiao_nrf52-s140/drivers/usb/usb_cdc.o (symbol from plugin): In function `USBD_IRQHandler':
(.text+0x0): multiple definition of `mp_hal_stdout_tx_strn_cooked'
build-seeed_xiao_nrf52-s140/drivers/bluetooth/ble_uart.o (symbol from plugin):(.text+0x0): first defined here
collect2: error: ld returned 1 exit status
make: *** [Makefile:518: build-seeed_xiao_nrf52-s140/firmware.elf] Error 1
```
I'm now completely stuck. There is really really few information concerning the uart_NUS, and absolutely no example to even start with. Using my sublime I did see ble_uart in the bluetooth drivers... so something must be there right? 

Do you/someone have some kind of hint/idea what I could do?
The circuit python NUS implementation has a separate "nordic.py" file handling the NUS. but lots of code inside is CP specific, so there isn't much I can extract from this... and at my level it's really not possible. ^^

**Comment by robert-hh** [comment]
Since MICROPY_HW_USB_CDC is defined in the board's mpconfigport.h, these functions are enabled twice. You can set MICROPY_HW_USB_CDC to 0, at which time there is not REPL any more over USB, but using UART and maybe Bluetooth.

**Reply by doh38** [reply]
thanks you. Well, it did compile, it did boot. but I couldn't find any ble_uart anywhere. And as soon as I'm really not comfortable with MP, I think I will let this project mature a bit if it ever will. 
For info, the usb-REPL is indeed inactive but just until the bluetooth connection is made (with my above connect-only script)... then works like before. I really have no idea :D
Specially with the "Only one uart device" the nrf is able to handle, BLE_UART, Uart from pin AND usb repl which is serial I think completely in the same time with the same dongle on CP or directly from C with the nordic sdk. 
I'm by any mean comparing MP and CP, just a bit frustrated to not being able to go pass this due to absoute lack of documentation. Thanks again for your help, I learn at least a few thing on the way and that's always worth it  :)

**Reply by robert-hh** [reply]
The lack of documentation is a problem, especially since the bluetooth API is different to the other boards. About enabling MICROPY_PY_BLE_NUS: If I do that, disable MICROPY_HW_USB_CDC and upload the created image to the board, I get a device called `mpus` created by the board without starting any other script, which is the bluetooth UART interface. I have no device or app ready which I could use for connecting.

**Reply by robert-hh** [reply]
Git it on my Mobile phone working., By just setting MICROPY_PY_BLE_NUS  (1) I get a REPL prompt at a bluetooth terminal. It is not required to start another script at the device. Only conneting has to be done shortly after reset.

**Reply by robert-hh** [reply]
Your script to use NUS works too WITHOUT enabling MICROPY_PY_BLE_NUS. Only one minor changes:
- the line testing the data has to be e.g. `led_red.high() if data[0] == 48 else led_red.low()`. Since data is a bytearray, data[0] is a number. And 48 is the ordinal value of the character '0'.

**Reply by doh38** [reply]
Hey Robert, you are pretty awesome :)) 
I will try this. 
I found out today that Nordic has changed their sdk strategy to Zephyr (in order to finally implement multi-core devices), and that's probably the reason "independant" devs are letting the "old" nrf on side. I will give your suggestion a try, I really hate to "lose" actually ;) Thanks again, have a nice day! let's let the snow melt a bit :)

**Reply by doh38** [reply]
And before I start. 
In CP, when initializing the ble_uart ( UARTService() ) we have to call it. something like:
```
``ble = BLERadio()
uart_BLE = UARTService()

 while ble.connected:
        rx_BLE_bytearray = uart_BLE.read(1) # add later crc and len check to adapt read length to received message

        led.value = False
        led2.value = True

        for data in rx_BLE_bytearray:
            uart_VESC.write(rx_BLE_bytearray) # Write to VESC
            print(rx_BLE_bytearray)
            led.value = True
            led2.value = False

        rx_VESC_bytearray = uart_VESC.read(1)

        for data in rx_VESC_bytearray:
            uart_BLE.write(rx_VESC_bytearray) #Write to BLE
            print("VESC", rx_VESC_bytearray)
            led.value = True
            led2.value = False
``````
So this function call (UARTService()) enable the uart.write and uart.read and handle the whole buffer story. since I didn't find anything similar that's where I was stuck (this and the wrong code where I thought it was simply not working).
If you have any idea how I could implement this that would be awesome. but as mentioned before, since my main project is working now on CP I may not spend too much effort on this and will wait that zephyr is correctly supported :)

**Reply by robert-hh** [reply]
If you enable MICROPY_PY_BLE_NUS, then you can use the sys.stdxx methods. sys.stdout.write() for writing and sys.stdin.read() for reading, or print() and input(). But I did not see a binding to a Python API.
If MICROPY_PY_BLE_NUS is not enabled, the ble_uart functions are not compiled in to the code.

**Reply by doh38** [reply]
thanks a ton, I really appreciate this :) and now I can at least have a few hints to start when I have digested my hard fail :)
